sdk/keyvault/test-resources-post.ps1

```diff
@@ -23,12 +23,84 @@ param (
     $RemainingArguments
 )
 
+# By default stop for any error.
+if (!$PSBoundParameters.ContainsKey('ErrorAction')) {
+    $ErrorActionPreference = 'Stop'
+}
+
 function Log($Message) {
     Write-Host ('{0} - {1}' -f [DateTime]::Now.ToLongTimeString(), $Message)
 }
 
-Write-Verbose "Example post-deployment script for Azure Key Vault in resource group '$ResourceGroupName' accessible to $TestApplicationOid."
-if ($DeploymentOutputs) {
-    $names = $DeploymentOutputs.Keys -join ', '
-    Log "Output variable names: $names"
+function New-X509Certificate2([string] $SubjectName) {
+
+    $rsa = [RSA]::Create(2048)
+    try {
+        $req = [CertificateRequest]::new(
+            [string] $SubjectName,
+            $rsa,
+            [HashAlgorithmName]::SHA256,
+            [RSASignaturePadding]::Pkcs1
+        )
+
+        # TODO: Add any KUs necessary to $req.CertificateExtensions
+
+        $NotBefore = [DateTimeOffset]::Now.AddDays(-1)
+        $NotAfter = $NotBefore.AddDays(365)
+
+        $req.CreateSelfSigned($NotBefore, $NotAfter)
+    }
+    finally {
+        $rsa.Dispose()
+    }
+}
+
+function Export-X509Certificate2([string] $Path, [X509Certificate2] $Certificate) {
+
+    $Certificate.Export([X509ContentType]::Pfx) | Set-Content $Path -AsByteStream
+}
+
+function Export-X509Certificate2PEM([string] $Path, [X509Certificate2] $Certificate) {
+
+@"
+-----BEGIN CERTIFICATE-----
+$([Convert]::ToBase64String($Certificate.RawData, 'InsertLineBreaks'))
+-----END CERTIFICATE-----
+"@ > $Path
+
+}
+
+# Make sure we deployed a Managed HSM.
+if (!$DeploymentOutputs['MANAGEDHSM_URL']) {
+    Log "Managed HSM not deployed; skipping activation"
+    exit
+}
+
+[Uri] $mhsmUrl = $DeploymentOutputs['MANAGEDHSM_URL']
+$mhsmName = $mhsmUrl.Host.Substring(0, $mhsmUrl.Host.IndexOf('.'))
+
+$tenant = $DeploymentOutputs['KEYVAULT_TENANT_ID']
+$username = $DeploymentOutputs['KEYVAULT_CLIENT_ID']
+$password = $DeploymentOutputs['KEYVAULT_CLIENT_SECRET']
+
+Log 'Creating 3 X509 certificates to activate security domain'
+$wrappingFiles = foreach ($i in 0..2) {
+    $certificate = New-X509Certificate2 "CN=$($mhsmUrl.Host)"
+
+    $baseName = "$PSScriptRoot\$mhsmName-certificate$i"
+    Export-X509Certificate2 "$baseName.pfx" $certificate
+    Export-X509Certificate2PEM "$baseName.cer" $certificate
+
+    Resolve-Path "$baseName.cer"
 }
+
+# TODO: Use Az module when available; for now, assumes Azure CLI is installed and in $Env:PATH.
+Log "Logging '$username' into the Azure CLI"
+az login --service-principal --tenant $tenant --username $username --password $password

```

Comment:
Would this logic be cached globally?

Reply:
Unfortunately, yes. But I'm hoping by the time we role this out across all languages (except, perhaps, Python) the new Az module will be ready. We will have to install it in CIs and assert it in the TestResourcs module, however, but I think that's better all around than potentially overwriting someone's azure-cli login. At least - for now - it's only limited to Key Vault.

/cc @weshaggard

Reply:
cc @mikeharder 

I wonder if there what the trade-offs are between using the Powershell modules vs using the CLI for our deployments. Is it easier and more reliable to obtain the cli then the powershell modules? If the cli also includes newer features sooner it that might be another reason to consider using cli.

Reply:
We already depend on PowerShell for cross-platform scripting. The only viable option available. Why require both PowerShell *and* Azure CLI when acquiring (or using, if already installed) the Az module has a built-in mechanism with supported PowerShell versions? Besides, the Azure CLI doesn't offer an option to disable saving credentials like Az does, which we use here to avoid overwriting the currently logged in principal - likely an interactive user; not really an issue either way for the CIs.